Source/DafnyCore/Compilers/CSharp/Compiler-Csharp.cs

```diff
@@ -265,27 +265,57 @@ public class CsharpCompiler : SinglePassCompiler {
+        }
+      }
 
-          if (typeParameters != null) {
-            var sep = "";
-            foreach (var ta in TypeArgumentInstantiation.ListFromFormals(typeParameters)) {
-              if (NeedsTypeDescriptor(ta.Formal)) {
-                var fieldName = FormatTypeDescriptorVariable(ta.Formal.GetCompileName(Options));
-                var decl = $"{DafnyTypeDescriptor}<{TypeName(ta.Actual, wTypeFields, ta.Formal.tok)}> {fieldName}";
-                wTypeFields.WriteLine($"private {decl};");
-                if (ta.Formal.Parent == cls) {
-                  wCtorParams.Write($"{sep}{decl}");
-                }
+      return new ClassWriter(this, wBody, wCtorBody);
+    }
 
-                wCtorBody.WriteLine($"this.{fieldName} = {TypeDescriptor(ta.Actual, wCtorBody, ta.Formal.tok)};");
-                sep = ", ";
-              }
+    /// <summary>
+    /// For each type parameter X in "typeParametersForClass" that needs a type descriptor,
+    ///   * Write "protected TypeDescriptor<X> _td_X;" to wTypeFields
+    ///     -- each entry is terminated by a newline
+    ///   * Write "TypeDescriptor<X> _td_X" to wCtorParams
+    ///     -- entries are separated by a comma
+    ///   * Write "_td_X" to wCallArguments
+    ///     -- entries are separated by a comma
+    ///   * Write "this._td_X := _td_X;" to wCtorBody
+    ///     -- each entry is terminated by a newline
+    /// Any of the writer parameters may be null, so long as at least one is non-null.
+    /// The method returns the number type descriptors written.
+    /// </summary>
+    int EmitTypeDescriptorsForClass(List<TypeParameter> typeParametersForClass, TopLevelDecl cls,
+      [CanBeNull] ConcreteSyntaxTree wTypeFields, [CanBeNull] ConcreteSyntaxTree wCtorParams,
+      [CanBeNull] ConcreteSyntaxTree wCallArguments, [CanBeNull] ConcreteSyntaxTree wCtorBody,
+      List<TypeParameter> alternateTypeParameters = null) {
+
+      var wError = wTypeFields ?? wCtorParams ?? wCallArguments ?? wCtorBody;
+      int numberOfEmittedTypeDescriptors = 0;
+      if (typeParametersForClass != null) {
+        var sep = "";
+        var j = 0;
+        foreach (var ta in TypeArgumentInstantiation.ListFromFormals(typeParametersForClass)) {
+          if (NeedsTypeDescriptor(ta.Formal)) {
+            var fieldName = FormatTypeDescriptorVariable((alternateTypeParameters == null ? ta.Formal : alternateTypeParameters[j]).GetCompileName(Options));
+            var actualType = alternateTypeParameters == null ? ta.Actual : new UserDefinedType(ta.Formal.tok, alternateTypeParameters[j]);
+            var paramName = TypeDescriptor(actualType, wError, ta.Formal.tok);
+            var decl = $"{DafnyTypeDescriptor}<{TypeName(actualType, wError, ta.Formal.tok)}> {fieldName}";
+
+            wTypeFields?.WriteLine($"protected {decl};");
+            if (ta.Formal.Parent == cls) {
+              wCtorParams?.Write($"{sep}{decl}");
```

Comment:
You can imagine a class `CommaSeparated` with a method `void Append(ICanRender element)` that automatically manages the `,` separator, so you replace:

```c#
        var wTypeFields = wr.Fork();
        var wCtorParams = new ConcreteSyntaxTree();
        var wCtorBody = wr.Format($"public {IdName(dt)}({wCtorParams})").NewBlock();
        EmitTypeDescriptorsForClass(dt.TypeArgs, dt, wTypeFields, wCtorParams, null, wCtorBody);
```

with

```c#
        var wTypeFields = wr.Fork();
        var wCtorParams = new CommaSeparated();
        var wCtorBody = wr.Format($"public {IdName(dt)}({wCtorParams})").NewBlock();
        EmitTypeDescriptorsForClass(dt.TypeArgs, dt, wTypeFields, wCtorParams, null, wCtorBody);
```

and 

```c#
int EmitTypeDescriptorsForClass(List<TypeParameter> typeParametersForClass, TopLevelDecl cls,
      [CanBeNull] ConcreteSyntaxTree wTypeFields, [CanBeNull] ConcreteSyntaxTree wCtorParams,
      [CanBeNull] ConcreteSyntaxTree wCallArguments, [CanBeNull] ConcreteSyntaxTree wCtorBody,
      List<TypeParameter> alternateTypeParameters = null) {
```

with 

```c#
int EmitTypeDescriptorsForClass(List<TypeParameter> typeParametersForClass, TopLevelDecl cls,
      [CanBeNull] ConcreteSyntaxTree wTypeFields, [CanBeNull] CommaSeparated wCtorParams,
      [CanBeNull] CommaSeparated wCallArguments, [CanBeNull] ConcreteSyntaxTree wCtorBody,
      List<TypeParameter> alternateTypeParameters = null) {
```

and then remove the variable `sep` in `EmitTypeDescriptorsForClass`.

And if you do, you can also consider adding an argument to `CommaSeparated` to manage to enclosing delimiters, in this case parenthesis, so you get

```c#
        var wTypeFields = wr.Fork();
        var wCtorParams = new CommaSeparated('(');
        var wCtorBody = wr.Format($"public {IdName(dt)}{wCtorParams}").NewBlock();
        EmitTypeDescriptorsForClass(dt.TypeArgs, dt, wTypeFields, wCtorParams, null, wCtorBody);
````

Reply:
I like this suggestion and I had thought about it when writing the code (because SO much of this code is spent getting commas and other delimiters in place!). However, it wasn't clear to me when `CommaSeparated` would look to see when to insert commas. That still isn't clear to me. If we could work this out, we'd save a lot of code in the compilers.